include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
+ *
+ *   * Cooperatively: when a thread explicitly calls #kthread_yield to
+ *     leave the CPU to other threads that might be ready at the same time.
+ *     Notice that, given the hard real-time guarantee, this only makes
+ *     sense if there are multiple ready threads at the same priority level,
+ *     as calling #kthread_yield will not schedule a lower-priority thread.
+ *     It is normally not necessary to call #thread_yield.
+ *
+ * In addition to threads (#kthread_t) and events (#kevent_t), the kernel
+ * library offers a message passing communication primitive called "message
+ * box" (#kmbox_t). A mailbox is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore. 
+ *
+ * Mailboxes can be useful as communicative primitives among threads
+ * (eg: to send tasks to other threads), or a synchronization primitive
+ * to wait for multiple events (see #kmbox_attach_event).
+ *
+ * @{
+ */
+
+/**
+ * @brief a kernel thread for parallel execution
+ *
+ * This structure represents a thread that can be scheduled
+ * for parallel execution. Create a thread with #kthread_new,
+ * allocating the required memory for the stack.
+ *
+ * Normally, there's no need to explicitly manage the lifetime
+ * or state of a thread. Once the thread is created, it is
+ * immediately started, and the kernel will scheduled it when
+ * required. If the thread exits (by simply returning by its
+ * entry point function), it will be cleaned up and the memory
+ * released.
+ *
+ * #kthread_new returns a pointer to the thread (a pointer
+ * to kthread_t) that can be used to externally manage the
+ * thread like changing its priority or killing it.
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
+	reg_block_t* stack_state;
+	/** Current state of interrupt depth for this thread */
+	int interrupt_depth;
+	/** Size of the stack in bytes */
+	int stack_size;
+	/** Name of thread (for debugging purposes) */
+	const char *name;
+	/** Internal flags */
```

Comment:
Are these for internal use only? Should we prefix with `__`?

Reply:
To tell you the truth, the whole `kthread_t` structure should be opaque from a formal API standpoint. User code should just manipulate "handles" to threads (so opaque pointers), and surely not mess with the thread structure contents. 

I resisted the temptation of hiding this in an internal header because in general I think that with libdragon we're generally better off keeping the door open to advanced hacking / tweaking of libraries, rather than trying to force a pure API that the users then need to hack around when it doesn't do everything possible. But I'm happy to be talked otherwise for the kernel. See also `timer_link_t` which is in `timer.h` but I fail to see good reasons for it, if not for a general "open" philosophy.

Reply:
I agree on ideally we should keep it internal. I'm not against keeping it open such that consumers can implement their own methods around them if need be. On the other hand, with an implementation so tightly coupled with the libdragon and its ex. vectors etc., I don't think it will be of much help.

We can still open the core helpers/mechanisms to the user (like we did for example with cop0/cop1 macros and custom ex handler registration etc.) to enable tweaking. As an example, RDP implementation can consist of many small pieces of helper funcs. / macros instead of a fully closed api like `rdp_enable_primitive_fill`. Rather it can implement the ring buffer and provide an API to enqueue commands. Then we could have a `enable_primitive_fill` "helper" instead, that we can enqueue.

A similar approach would not probably be easy at the exception/kernel level though. Instead we may try adding a "hook" where we actually schedule the threads, such that third party code can implements its own "kernel extensions" but that would potentially bring the performance down and would be a mess in practice. Instead, one should provide her/his own `inthandler` and kernel modules if s/he wants to tweak it. I think we should aim for this kind of an architecture for long term expandibility/tweaking. This may not be easy considering how little time we can spare for this project (except for @rasky :)) though.

In short, I think it is ok to keep it open but I see little value doing so for such a highly tightly coupled portion of the library.

Reply:
OK I agree with you. Let's keep it this way.